v3/mathjax3-ts/input/mathml.ts

```diff
@@ -0,0 +1,169 @@
+        this.parser = this.options['DOMParser'] || new (DOM.DOMParser)();
+        this.mmlFilters = new FunctionList();
+    }
+
+    /*
+     * Don't process strings (process nodes)
+     *
+     * @override
+     */
+    public get processStrings() {
+        return false;
+    }
+
+    /*
+     * Convert a MathItem to internal format:
+     *   If there is no existing MathML node, or we are asked to reparse everything
+     *     Execute the preFilters on the math
+     *     Parse the MathML string in the desired format, and check the result for errors
+     *     If we got an HTML document:
+     *       Check that it has only one child (the <math> element), and use it
+     *     Otherwise
+     *       Use the root element from the XML document
+     *     If the node is not a <math> node, report the error.
+     *   Execute the mmlFilters on teh parsed MathML
+     *   Compile the MathML to internal format, and execute the postFilters
+     *   Return the resulting internal format
+     *
+     * @override
+     */
+    public compile(math: MathItem) {
+        let mml = math.start.node;
+        if (!mml || this.options['forceReparse']) {
+            let mathml = this.executeFilters(this.preFilters, math, math.math || '<math></math>');
+            let doc = this.parser.parseFromString(mathml, 'text/' + this.options['parseAs']);
+            doc = this.checkForErrors(doc);
+            if (doc.body) {
+                if (doc.body.childNodes.length !== 1) {
+                    this.error('MathML must consist of a single element');
+                }
+                mml = doc.body.removeChild(doc.body.firstChild) as Element;
+            } else {
+                mml = doc.removeChild(doc.firstChild) as Element;
+            }
+            if (mml.nodeName.toLowerCase().replace(/^[a-z]+:/, '') !== 'math') {
+                this.error('MathML must be formed by a <math> element, not <' +
+                           mml.nodeName.toLowerCase() + '>');
+            }
+        }
+        mml = this.executeFilters(this.mmlFilters, math, mml);
+        return this.executeFilters(this.postFilters, math, this.mathml.compile(mml as HTMLElement));
+    }
+
+    /*
+     * Check a parsed MathML string for errors.
+     *
+     * @param{Document} doc  The document returns from the DOMParser
+     * @return{Document}     The document
+     */
+    protected checkForErrors(doc: Document) {
+        let err = doc.querySelector('parsererror');
+        if (err) {
+            if (err.textContent === '') {
+                this.error('Error processing MathML');
+            }
+            this.options['parseError'].call(this, err);
+        }
+        return doc;
+    }
+
+    /*
+     * Throw an error
+     *
+     * @param{string} message  The error message to produce
+     */
+    protected error(message: string) {
+        // FIXME:  should this be creating merror nodes instead?
+        throw new Error(message);
```

Comment:
Not always.  Some calls are not because it didn't parse, but because of other errors (like the root node isn't a `<math>` node).